File: backend/app/Services/Bintex/Yes1Reader.php
<?php

declare(strict_types=1);

namespace App\Services\Bintex;

use RuntimeException;

/**
 * PHP port of yuku.alkitab.yes1.Yes1Reader — reads YES1 (legacy) Bible files.
 *
 * YES1 header: 0x98 0x58 0x0D 0x0A 0x00 0x5D 0xE0 0x01
 * Sequential sections with 12-byte padded names (underscore-padded).
 * Section format: [12-byte name][4-byte big-endian size][data...]
 * Verse text: newline (0x0a) separated within chapter blocks.
 * No xrefs or footnotes support.
 */
class Yes1Reader
{
    private const HEADER = "\x98\x58\x0D\x0A\x00\x5D\xE0\x01";
    private const HEADER_LENGTH = 8;
    private const SECTION_NAME_LENGTH = 12;

    private string $data;
    private int $length;

    // Parsed info
    private ?array $versionInfo = null;
    private ?array $booksInfo = null;
    private ?int $textBaseOffset = null;

    public function __construct(string $data)
    {
        $this->data = $data;
        $this->length = strlen($data);
        $this->parseHeader();
    }

    public static function fromFile(string $path): self
    {
        $data = file_get_contents($path);
        if ($data === false) {
            throw new RuntimeException("Cannot read file: {$path}");
        }
        return new self($data);
    }

    private function parseHeader(): void
    {
        if ($this->length < self::HEADER_LENGTH) {
            throw new RuntimeException('YES1: File too small for header');
        }
        $header = substr($this->data, 0, self::HEADER_LENGTH);
        if ($header !== self::HEADER) {
            throw new RuntimeException('YES1: Invalid header');
        }
    }

    /**
     * Find a section by name. Scans sequentially from offset 8.
     *
     * @return array{offset: int, size: int}|null
     */
    private function findSection(string $sectionName): ?array
    {
        // Pad section name to 12 bytes with underscores
        $paddedName = str_pad($sectionName, self::SECTION_NAME_LENGTH, '_');
        $pos = self::HEADER_LENGTH;

        while ($pos + self::SECTION_NAME_LENGTH + 4 <= $this->length) {
            $name = substr($this->data, $pos, self::SECTION_NAME_LENGTH);
            $pos += self::SECTION_NAME_LENGTH;

            // Read 4-byte big-endian size
            $size = unpack('N', substr($this->data, $pos, 4))[1];
            if ($size >= 0x80000000) {
                $size -= 0x100000000;
            }
            $pos += 4;

            if ($name === '____________') {
                // End marker
                return null;
            }

            if ($name === $paddedName) {
                return ['offset' => $pos, 'size' => $size];
            }

            // Skip this section's data
            $pos += $size;
        }

        return null;
    }

    // -- Version Info --

    public function getVersionInfo(): array
    {
        if ($this->versionInfo !== null) {
            return $this->versionInfo;
        }

        $section = $this->findSection('infoEdisi');
        if ($section === null) {
            throw new RuntimeException('YES1: infoEdisi section not found');
        }

        $sectionData = substr($this->data, $section['offset'], $section['size']);
        $br = new BintexReader($sectionData);

        $info = [
            'shortName' => null,
            'longName' => null,
            'description' => null,
            'locale' => null,
            'book_count' => 0,
            'hasPericopes' => 0,
            'textEncoding' => 1,
        ];

        while (true) {
            $key = $br->readShortString();

            switch ($key) {
                case 'versi':
                case 'format':
                    $br->readInt();
                    break;
                case 'nama':
                    $br->readShortString(); // internal name, not used
                    break;
                case 'shortName':
                case 'shortTitle':
                    $info['shortName'] = $br->readShortString();
                    break;
                case 'judul':
                    $info['longName'] = $br->readShortString();
                    break;
                case 'keterangan':
                    $info['description'] = $br->readLongString();
                    break;
                case 'nkitab':
                    $info['book_count'] = $br->readInt();
                    break;
                case 'perikopAda':
                    $info['hasPericopes'] = $br->readInt();
                    break;
                case 'encoding':
                    $info['textEncoding'] = $br->readInt();
                    break;
                case 'locale':
                    $info['locale'] = $br->readShortString();
                    break;
                case 'end':
                    break 2;
                default:
                    // Unknown key — skip. In original Java this throws; we're more lenient.
                    break 2;
            }
        }

        $this->versionInfo = $info;
        return $this->versionInfo;
    }

    // -- Books Info --

    /**
     * @return array<int, array{bookId: int, shortName: ?string, offset: int, chapter_count: int, verse_counts: int[], chapter_offsets: int[]}>
     */
    public function getBooksInfo(): array
    {
        if ($this->booksInfo !== null) {
            return $this->booksInfo;
        }

        $versionInfo = $this->getVersionInfo();
        $bookCount = $versionInfo['book_count'];

        $section = $this->findSection('infoKitab');
        if ($section === null) {
            throw new RuntimeException('YES1: infoKitab section not found');
        }

        $sectionData = substr($this->data, $section['offset'], $section['size']);
        $br = new BintexReader($sectionData);

        $books = [];
        for ($bookIndex = 0; $bookIndex < $bookCount; $bookIndex++) {
            $book = [
                'bookId' => -1,
                'shortName' => null,
                'abbreviation' => null,
                'offset' => 0,
                'chapter_count' => 0,
                'verse_counts' => [],
                'chapter_offsets' => [],
            ];
            $empty = false;

            for ($keyIndex = 0; ; $keyIndex++) {
                $key = $br->readShortString();

                switch ($key) {
                    case 'versi':
                        $br->readInt();
                        break;
                    case 'pos':
                        $book['bookId'] = $br->readInt();
                        break;
                    case 'nama':
                    case 'judul':
                        $book['shortName'] = $br->readShortString();
                        break;
                    case 'npasal':
                        $book['chapter_count'] = $br->readInt();
                        break;
                    case 'nayat':
                        $verseCounts = [];
                        for ($i = 0; $i < $book['chapter_count']; $i++) {
                            $verseCounts[] = $br->readUint8();
                        }
                        $book['verse_counts'] = $verseCounts;
                        break;
                    case 'ayatLoncat':
                    case 'pdbBookNumber':
                        $br->readInt(); // ignored
                        break;
                    case 'pasal_offset':
                        // YES1 has chapter_count + 1 offsets
                        $offsets = [];
                        for ($i = 0; $i <= $book['chapter_count']; $i++) {
                            $offsets[] = $br->readInt();
                        }
                        $book['chapter_offsets'] = $offsets;
                        break;
                    case 'encoding':
                        $br->readInt(); // ignored, deprecated
                        break;
                    case 'offset':
                        $book['offset'] = $br->readInt();
                        break;
                    case 'end':
                        if ($keyIndex === 0) {
                            $empty = true;
                        }
                        break 2;
                    default:
                        break 2;
                }
            }

            if (!$empty && $book['bookId'] >= 0) {
                $books[$book['bookId']] = $book;
            }
        }

        $this->booksInfo = $books;
        return $this->booksInfo;
    }

    // -- Text --

    private function getTextBaseOffset(): int
    {
        if ($this->textBaseOffset !== null) {
            return $this->textBaseOffset;
        }

        $section = $this->findSection('teks');
        if ($section === null) {
            throw new RuntimeException('YES1: teks section not found');
        }

        $this->textBaseOffset = $section['offset'];
        return $this->textBaseOffset;
    }

    /**
     * Load verse texts for a specific book and chapter.
     * YES1 stores chapter text as newline-separated (0x0a) verse blocks.
     *
     * @return string[]
     */
    public function loadVerseText(int $bookId, int $chapter1): array
    {
        $books = $this->getBooksInfo();
        if (!isset($books[$bookId])) {
            throw new RuntimeException("YES1: Book ID {$bookId} not found");
        }

        $book = $books[$bookId];
        if ($chapter1 <= 0 || $chapter1 > $book['chapter_count']) {
            throw new RuntimeException("YES1: Chapter {$chapter1} out of range for book {$bookId}");
        }

        $versionInfo = $this->getVersionInfo();
        $textBaseOffset = $this->getTextBaseOffset();

        $seekTo = $textBaseOffset + $book['offset'] + $book['chapter_offsets'][$chapter1 - 1];
        $chapterLength = $book['chapter_offsets'][$chapter1] - $book['chapter_offsets'][$chapter1 - 1];

        if ($chapterLength <= 0) {
            return [];
        }

        $raw = substr($this->data, $seekTo, $chapterLength);

        // Split by newline (0x0a)
        $parts = explode("\n", $raw);

        // Remove trailing empty string if any
        if (count($parts) > 0 && $parts[count($parts) - 1] === '') {
            array_pop($parts);
        }

        // Convert encoding if needed
        if ($versionInfo['textEncoding'] === 1) {
            // ASCII / Latin-1
            return array_map(fn ($v) => mb_convert_encoding($v, 'UTF-8', 'ISO-8859-1'), $parts);
        }

        // UTF-8 (encoding == 2)
        return $parts;
    }

    // -- Pericopes --

    /**
     * Load pericope data from perikopIndex and perikopBlok sections.
     *
     * @return array<int, array{ari: int, title: ?string, parallels: string[]}>
     */
    public function loadPericopes(): array
    {
        $versionInfo = $this->getVersionInfo();
        if (($versionInfo['hasPericopes'] ?? 0) === 0) {
            return [];
        }

        $indexSection = $this->findSection('perikopIndex');
        if ($indexSection === null) {
            return [];
        }

        $indexData = substr($this->data, $indexSection['offset'], $indexSection['size']);
        $br = new BintexReader($indexData);

        // Read pericope index
        $version = $br->readUint8();
        $entryCount = $br->readInt();

        $aris = [];
        $offsets = [];
        for ($i = 0; $i < $entryCount; $i++) {
            $aris[] = $br->readInt();
            $offsets[] = $br->readInt();
        }

        // Read pericope blocks
        $blockSection = $this->findSection('perikopBlok');
        if ($blockSection === null) {
            return [];
        }

        $blockData = substr($this->data, $blockSection['offset'], $blockSection['size']);
        $result = [];

        for ($i = 0; $i < $entryCount; $i++) {
            $bbr = new BintexReader($blockData, $offsets[$i]);

            $blockVersion = $bbr->readUint8();
            $title = $bbr->readValueString();
            $parallelCount = $bbr->readUint8();
            $parallels = [];
            for ($j = 0; $j < $parallelCount; $j++) {
                $parallels[] = $bbr->readValueString();
            }

            $result[] = [
                'ari' => $aris[$i],
                'title' => $title,
                'parallels' => $parallels,
            ];
        }

        return $result;
    }

    /**
     * Get pericopes for a specific chapter.
     *
     * @return array<int, array{ari: int, title: ?string, parallels: string[]}>
     */
    public function getPericopesForChapter(int $bookId, int $chapter1): array
    {
        $allPericopes = $this->loadPericopes();
        $result = [];

        foreach ($allPericopes as $pericope) {
            $ari = $pericope['ari'];
            $periBookId = ($ari >> 16) & 0xff;
            $periChapter = ($ari >> 8) & 0xff;

            if ($periBookId === $bookId && $periChapter === $chapter1) {
                $result[] = $pericope;
            }
        }

        return $result;
    }

    // YES1 does not support xrefs or footnotes
    public function getXrefEntry(int $arif): ?string
    {
        return null;
    }

    public function getFootnoteEntry(int $arif): ?string
    {
        return null;
    }
}

File: mobile/composeApp/src/commonMain/kotlin/org/androidbible/data/bintex/Yes1Reader.kt
package org.androidbible.data.bintex

/**
 * Kotlin port: reads YES1 (legacy) Bible files.
 *
 * YES1 header: 0x98 0x58 0x0D 0x0A 0x00 0x5D 0xE0 0x01
 * Sequential sections with 12-byte underscore-padded names.
 * Verse text: newline (0x0A) separated within chapter blocks.
 */
class Yes1Reader(private val data: ByteArray) {

    private var cachedVersionInfo: Yes2VersionInfo? = null
    private var cachedBooksInfo: Map<Int, Yes2Book>? = null
    private var textBaseOffset: Int? = null

    init {
        check(data.size >= HEADER.size) { "YES1: File too small for header" }
        for (i in HEADER.indices) {
            check(data[i] == HEADER[i]) { "YES1: Invalid header at byte $i" }
        }
    }

    /** Find a section by name. Scans sequentially from offset 8. */
    private fun findSection(sectionName: String): SectionLocation? {
        val paddedName = sectionName.padEnd(SECTION_NAME_LENGTH, '_')
        var pos = HEADER.size

        while (pos + SECTION_NAME_LENGTH + 4 <= data.size) {
            val nameBytes = data.copyOfRange(pos, pos + SECTION_NAME_LENGTH)
            val name = String(CharArray(nameBytes.size) { (nameBytes[it].toInt() and 0xFF).toChar() })
            pos += SECTION_NAME_LENGTH

            val size = readInt32At(pos)
            pos += 4

            if (name == "____________") return null // end marker
            if (name == paddedName) return SectionLocation(pos, size)

            pos += size
        }
        return null
    }

    private fun readInt32At(offset: Int): Int {
        return ((data[offset].toInt() and 0xFF) shl 24) or
            ((data[offset + 1].toInt() and 0xFF) shl 16) or
            ((data[offset + 2].toInt() and 0xFF) shl 8) or
            (data[offset + 3].toInt() and 0xFF)
    }

    fun getVersionInfo(): Yes2VersionInfo {
        cachedVersionInfo?.let { return it }

        val section = findSection("infoEdisi") ?: error("YES1: infoEdisi section not found")
        val sectionData = data.copyOfRange(section.offset, section.offset + section.size)
        val br = BintexReader(sectionData)

        var shortName: String? = null
        var longName: String? = null
        var description: String? = null
        var locale: String? = null
        var bookCount = 0
        var hasPericopes = false
        var textEncoding = 1

        while (br.remaining > 0) {
            val key = br.readShortString()
            when (key) {
                "versi", "format" -> br.readInt()
                "nama" -> br.readShortString()
                "shortName", "shortTitle" -> shortName = br.readShortString()
                "judul" -> longName = br.readShortString()
                "keterangan" -> description = br.readLongString()
                "nkitab" -> bookCount = br.readInt()
                "perikopAda" -> hasPericopes = br.readInt() != 0
                "encoding" -> textEncoding = br.readInt()
                "locale" -> locale = br.readShortString()
                "end" -> break
                else -> break
            }
        }

        val info = Yes2VersionInfo(shortName, longName, description, locale, bookCount, hasPericopes, textEncoding)
        cachedVersionInfo = info
        return info
    }

    fun getBooksInfo(): Map<Int, Yes2Book> {
        cachedBooksInfo?.let { return it }

        val versionInfo = getVersionInfo()
        val section = findSection("infoKitab") ?: error("YES1: infoKitab section not found")
        val sectionData = data.copyOfRange(section.offset, section.offset + section.size)
        val br = BintexReader(sectionData)

        val books = LinkedHashMap<Int, Yes2Book>()

        for (bookIndex in 0 until versionInfo.bookCount) {
            var bookId = -1
            var shortName: String? = null
            var chapterCount = 0
            var verseCounts = intArrayOf()
            var chapterOffsets = intArrayOf()
            var offset = 0
            var empty = false

            var keyIndex = 0
            while (br.remaining > 0) {
                val key = br.readShortString()
                when (key) {
                    "versi" -> br.readInt()
                    "pos" -> bookId = br.readInt()
                    "nama", "judul" -> shortName = br.readShortString()
                    "npasal" -> chapterCount = br.readInt()
                    "nayat" -> verseCounts = IntArray(chapterCount) { br.readUint8() }
                    "ayatLoncat", "pdbBookNumber" -> br.readInt()
                    "pasal_offset" -> chapterOffsets = IntArray(chapterCount + 1) { br.readInt() }
                    "encoding" -> br.readInt()
                    "offset" -> offset = br.readInt()
                    "end" -> {
                        if (keyIndex == 0) empty = true
                        break
                    }
                    else -> break
                }
                keyIndex++
            }

            if (!empty && bookId >= 0) {
                books[bookId] = Yes2Book(bookId, shortName, null, offset, chapterCount, verseCounts, chapterOffsets)
            }
        }

        cachedBooksInfo = books
        return books
    }

    private fun getTextBaseOffset(): Int {
        textBaseOffset?.let { return it }
        val section = findSection("teks") ?: error("YES1: teks section not found")
        textBaseOffset = section.offset
        return section.offset
    }

    /**
     * Load verse texts for a specific bookId and 1-based chapter.
     * YES1 stores chapter text as newline-separated (0x0A) verse blocks.
     */
    fun loadVerseText(bookId: Int, chapter1: Int): List<String> {
        val books = getBooksInfo()
        val book = books[bookId] ?: error("YES1: Book ID $bookId not found")
        check(chapter1 in 1..book.chapterCount) { "YES1: Chapter $chapter1 out of range for book $bookId" }

        val versionInfo = getVersionInfo()
        val base = getTextBaseOffset()

        val seekTo = base + book.offset + book.chapterOffsets[chapter1 - 1]
        val chapterLength = book.chapterOffsets[chapter1] - book.chapterOffsets[chapter1 - 1]

        if (chapterLength <= 0) return emptyList()

        val raw = data.copyOfRange(seekTo, seekTo + chapterLength)
        val text = if (versionInfo.textEncoding == 1) {
            String(CharArray(raw.size) { (raw[it].toInt() and 0xFF).toChar() })
        } else {
            raw.decodeToString()
        }

        val parts = text.split('\n').toMutableList()
        if (parts.isNotEmpty() && parts.last().isEmpty()) {
            parts.removeAt(parts.lastIndex)
        }
        return parts
    }

    /** Read YES1 helper: short string (1-byte length prefix, UTF-16BE chars). */
    private fun BintexReader.readShortString(): String {
        val len = readUint8()
        if (len == 0) return ""
        return String(CharArray(len) { readUint16().toChar() })
    }

    /** Read YES1 helper: long string (4-byte length prefix, UTF-16BE chars). */
    private fun BintexReader.readLongString(): String {
        val len = readInt()
        if (len == 0) return ""
        return String(CharArray(len) { readUint16().toChar() })
    }

    private data class SectionLocation(val offset: Int, val size: Int)

    companion object {
        private val HEADER = byteArrayOf(
            0x98.toByte(), 0x58, 0x0D, 0x0A, 0x00, 0x5D, 0xE0.toByte(), 0x01
        )
        private const val SECTION_NAME_LENGTH = 12

        fun fromByteArray(data: ByteArray): Yes1Reader = Yes1Reader(data)
    }
}
